Add JSONAPI Errors

dimroc opened the pull request.

rupurt commented on store/presenters/presenters.go:
```diff
@@ -142,8 +144,10 @@ func NewConfigWhitelist(config store.Config) ConfigWhitelist {
-		RootDir: config.RootDir,
+		Port:             config.Port,
+		RootDir:          config.RootDir,
+		SessionTimeout:   config.SessionTimeout,
```
Are `SessionTimeout` & `ReaperExpiration` from another PR?

dimroc replied:
Yes, but they weren't added to presenters.

dimroc commented on gui/__tests__/containers/SignIn.test.js:
```diff
@@ -59,8 +59,8 @@ describe('containers/SignIn', () => {
     const newState = store.getState()
-    expect(newState.errors).toEqual({
-      messages: ['Your email or password is incorrect. Please try again'],
+    expect(newState.routing).toEqual({
```
In this initial implementation I felt the need to have errors in each state group because global errors kept rerendering the whole Layout and clobbering the input textarea.

What initially happened
1. `Layout.js` listens to global errors.
2.  Submitting `Job/BridgeForm` adds this new unauthorized, sign back in error.
3. Triggers a rerender of layout where it realizes its authentication.allowed is false
4. Renders sign in page.

What happens now
1. `Layout.js` only listens to `routing.errors` (we can rename this) and works as expected
2. `Job/BridgeForm` only listens to `create.errors`
3.  Submitting `Job/BridgeForm` adds this new unauthorized, sign back in error.
4. `Job/BridgeForm` only rerenders `FormNotifications` and doesn't clobber input textarea. 

Potential solution
1. `Layout.js` listens to global errors.
2.  Submitting `Job/BridgeForm` adds this new unauthorized, sign back in error.
3. We keep authentication.allowed true to prevent redirect to sign i page.
4. Rerender of Layout keeps content.

I'm open to other ideas. I'm going to think on this a bit myself.

dimroc commented on gui/src/components/FormNotifications.js:
```diff
@@ -1,43 +1,66 @@
+            </Flash>
+          )}
+        {!authenticated && (
+          <Flash warning className={classes.flash}>
```
This was just copied and pasted from the previous implementation. I'll take a deeper look.

dimroc commented on gui/src/errors.js:
```diff
@@ -1,5 +1,8 @@
-export function AuthenticationError (message) {
-  this.message = message
+export function AuthenticationError (response) {
+  this.errors = [{
```
This is so our AuthenticationError complies with JSONAPI Error and is consistent in our app.

rupurt replied:
How does `this.errors` get used? From my understanding it will be a private attribute that never gets used.

dimroc replied:
When we 401, we throw this error:
https://github.com/smartcontractkit/chainlink/blob/features/gui-error-message-on-session-expiration/gui/src/api.js#L35-L37

Where it is then consumed by the `reducer/notifications`:
https://github.com/smartcontractkit/chainlink/blob/features/gui-error-message-on-session-expiration/gui/src/connectors/redux/reducers/notifications.js#L38

dimroc commented:
Made your requested changes. Could use some feedback @rupurt on how to get test coverage on the error reporting in `Layout.js` since that top level component doesn't have a matching test file, but the subcomponents do. Break error reporting into its component?

dimroc commented on gui/src/components/FormNotifications.js:
```diff
@@ -2,42 +2,27 @@ import React, { Fragment } from 'react'
+    return (
+      <Fragment>
+        {JSON.stringify(success) !== '{}' && (
+          <Flash success className={classes.flash}>
```
Note: Error reporting has been removed from this class.

@rupurt @NavyAdmiral Now that we've placed `create` error reporting at the top level in `Layout.js` and this only notifies successes, I believe we should ultimately move this to `Layout.js` (or top level) as well.

In other words, all notifications, successes and errors, should be at the top level and not embedded in contextual components. No need to do that in this PR but that's where we are moving. It will also get rid of the `jobOrBridge` but we'll have to come up with a way to generate the show link.

Top level error reporting: https://github.com/smartcontractkit/chainlink/pull/537/files#diff-4c5a8365a1292c74b8df56f96414fc09R183

NavyAdmiral replied:
@dimroc reporting errors in form components was a design choice I made because I wanted to display Flash notifications within the `<TabContainer/>`. It will be better if we move it to top because the former approach results in complications for other notifications we might need in future outside these forms. Also, `jobOrBridge` was very convoluted and I very much disliked it. :shipit:

dimroc commented:
Ok @rupurt and @NavyAdmiral, I've renamed the `errors` reducer to `notifications` and renamed `BridgeAndJobFormNotifications` to just `Notifications`. This new reducer and component should consolidate the handling of both error and success notifications. I'm pretty happy with it. Let me know what you think.

NavyAdmiral commented on gui/src/components/Notifications.js:
```diff
@@ -0,0 +1,66 @@
+  }
+}
+
+const successPresenter = (success, i) => {
```
why is errorPresenter within the class while successPreenter is not?

dimroc replied:
Because `errorPresenter` uses `this.signOutLocally`.

NavyAdmiral replied:
why can't successPresenter go inside? 🗡

dimroc replied:
It can.

NavyAdmiral commented:
@dimroc I like current approach. Should make notifications for #543 easier to implement. 🥇

rupurt commented on gui/src/Layout.js:
```diff
@@ -177,12 +171,7 @@ class Layout extends Component {
-                  {errors.map((msg, i) => <p key={i}>{msg}</p>)}
-                </Flash>
-              }
+              <Notifications classes={classes} />
```
Big fan of making this consistent for error + success messages in the layout

rupurt commented on gui/src/components/LinkButton.js:
```diff
@@ -0,0 +1,6 @@
+import React from 'react'
+import StyledButton from '@material-ui/core/Button'
+
+export default props => (
```
I feel like this should allow classes to be assigned for styling. How do you feel about making the `SIGN IN AGAIN` link white underline title case?

rupurt commented on gui/src/connectors/redux/reducers/notifications.js:
```diff
@@ -26,7 +29,19 @@ export default (state = initialState, action = {}) => {
-        {messages: [SIGN_IN_FAIL_MSG]}
+        {errors: [{detail: SIGN_IN_FAIL_MSG}]}
+      )
+    case RECEIVE_CREATE_ERROR:
```
How do you feel about `*_ERROR` actions clearing `*_SUCCESS` actions an vice versa? Currently if you successfully create something and then get an error it shows both the success + error message.

<img width="1043" alt="screen shot 2018-09-05 at 8 19 27 pm" src="https://user-images.githubusercontent.com/680789/45133375-ad1eb580-b149-11e8-843b-34a40d31639a.png">
